File: C_plus_plus/BaseTreapNode.cpp
#pragma once

#include <stdio.h>
#include <stdlib.h>

#define MAX_TREAP_NODES 10000000

class BaseTreapNode
{
private:
	int priority;
	BaseTreapNode* left;
	BaseTreapNode* right;
	int value;
	int count;

public:

    BaseTreapNode() {
    }
    
	BaseTreapNode(int value)
	{
		priority = rand();
		this->value = value;
		left = NULL;
		right = NULL;
		count = 1;
	}
	
	BaseTreapNode* get_left()
	{
		return left;
	}

	virtual void set_left(BaseTreapNode* node)
	{
		left = node;
		update();
	}

	BaseTreapNode* get_right()
	{
		return right;
	}

	virtual void set_right(BaseTreapNode* node)
	{
		right = node;
		update();
	}

	int get_value()
	{
		return value;
	}

	void set_value(int value)
	{
		this->value = value;
	}

	int get_count()
	{
		return count;
	}

	int get_priority()
	{
		return priority;
	}

	virtual BaseTreapNode* rotateLeft()
	{
		BaseTreapNode* temp  = right;
		set_right(right->left);
		temp->set_left(this);
		return temp;
	}
	
	virtual BaseTreapNode* rotateRight()
	{
		BaseTreapNode* temp  = left;
		set_left(left->right);
		temp->set_right(this);
		return temp;
	}


    BaseTreapNode* deleteRoot()
	{
		BaseTreapNode* temp;

		if(left == NULL)
		{
			return right;
		}

		if(right == NULL)
		{
			return left;
		}

		if(left->priority < right->priority)
		{
			temp        = rotateRight();
			temp->right  = deleteRoot();
		}
		else
		{
			temp        = rotateLeft();
			temp->left   = deleteRoot();
		}
		return temp;
	}

	virtual void updateAggregate()
	{
	}

protected:
	virtual void update()
	{
		update_count();
	}


private: 
	void update_count()
	{
		count = (left != NULL ? left->count : 0) + 1 + (right != NULL ? right->count : 0);
	}

};

File: C_Sharp/Treap/BaseTreapNode.cs
using System;

namespace Treap
{
    [Serializable]
    public class BaseTreapNode<T> 
    {
        protected static readonly Random RndPriority = new Random(42);

        private int count;
        private T value;
        private readonly int priority;
        private BaseTreapNode<T> left;
        private BaseTreapNode<T> right;

        public BaseTreapNode(T value)
        {
            this.value = value;
            priority = RndPriority.Next();
            count = 1;
        }

        public T Value
        {
            get
            {
                return value;
            }
            set
            {
                this.value = value;

            }
        }

        public int Priority
        {
            get
            {
                return priority;
            }
			
        }

        public BaseTreapNode<T> Left
        {
            get
            {
                return left;
            }
			
            set { LeftSetter(value); }
        }

        protected virtual void LeftSetter(BaseTreapNode<T> node)
        {
            left = node;
            Update();
        }

        public BaseTreapNode<T> Right
        {
            get
            {
                return right;
            }
			
            set { RightSetter(value); }
        }

        protected virtual void RightSetter(BaseTreapNode<T> node)
        {
            right = node;
            Update();
        }

        public int Count
        {
            get { return count; }
        }

        protected virtual void Update()
        {
            UpdateCount();
        }

        private void UpdateCount()
        {
            count = (left != null ? left.count : 0) + 1 + (right != null ? right.count : 0);
        }


        public virtual BaseTreapNode<T> RotateLeft()
        {
            BaseTreapNode<T> temp  = Right;
            Right           = Right.Left;
            temp.Left       = this;
            return temp;
        }

        public virtual BaseTreapNode<T> RotateRight()
        {
            BaseTreapNode<T> temp  = Left;
            Left            = Left.Right;
            temp.Right      = this;
            return temp;
        }

        public BaseTreapNode<T> DeleteRoot()
        {
            BaseTreapNode<T> temp;
			
            if(Left == null)
            {
                return Right;
            }
			
            if(Right == null)
            {
                return Left;
            }
			
            if(Left.Priority < Right.Priority)
            {
                temp        = RotateRight();
                temp.Right  = DeleteRoot();
            }
            else
            {
                temp        = RotateLeft();
                temp.Left   = DeleteRoot();
            }

            return temp;
        }

        public virtual void UpdateAggregate()
        {
        }
    }
}